java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowFileReader.java

```diff
@@ -164,12 +146,59 @@ public boolean loadNextBatch() throws IOException {
       ArrowBlock block = footer.getRecordBatches().get(currentRecordBatch++);
       ArrowRecordBatch batch = readRecordBatch(in, block, allocator);
       loadRecordBatch(batch);
+      try {
+        loadDictionaries();
+      } catch (IOException e) {
+        batch.close();
+        throw e;
+      }
       return true;
     } else {
       return false;
     }
   }
 
+  /**
+   * Loads any dictionaries that may be needed by the given record batch. It attempts
+   * to read as little as possible but may read in more deltas than are necessary for blocks
+   * toward the end of the file.
+   */
+  private void loadDictionaries() throws IOException {
+    // initial load
+    if (currentDictionaryBatch == 0) {
+      for (int i = 0; i < dictionaries.size(); i++) {
+        ArrowBlock block = footer.getDictionaries().get(currentDictionaryBatch++);
+        ArrowDictionaryBatch dictionaryBatch = readDictionaryBatch(in, block, allocator);
+        loadDictionary(dictionaryBatch, false);
+      }
+      estimatedDictionaryRecordBatch++;
+    } else {
+      // we need to look for delta dictionaries. It involves a look-ahead, unfortunately.
+      HashSet<Long> visited = new HashSet<Long>();
+      while (estimatedDictionaryRecordBatch < currentRecordBatch &&
+             currentDictionaryBatch < footer.getDictionaries().size()) {
+        ArrowBlock block = footer.getDictionaries().get(currentDictionaryBatch++);
+        ArrowDictionaryBatch dictionaryBatch = readDictionaryBatch(in, block, allocator);
+        long dictionaryId = dictionaryBatch.getDictionaryId();
+        if (visited.contains(dictionaryId)) {
```

Comment:
Condition 'visited.contains(dictionaryId)' is always 'false'

Reply:
Doh, left over cruft. thanks.